tango/workspaces/remote_workspace.py

```diff
@@ -0,0 +1,392 @@
+import json
+import logging
+import random
+import tempfile
+import warnings
+from abc import abstractmethod
+from collections import OrderedDict
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Dict, Generator, Iterable, Iterator, Optional, TypeVar, Union, cast
+from urllib.parse import ParseResult
+
+import petname
+
+from tango.common.exceptions import StepStateError
+from tango.common.logging import file_handler
+from tango.common.remote_utils import (
+    RemoteClient,
+    RemoteConstants,
+    RemoteDataset,
+    RemoteDatasetConflict,
+    RemoteDatasetNotFound,
+    RemoteStepLock,
+)
+from tango.common.util import exception_to_string, tango_cache_dir, utc_now_datetime
+from tango.step import Step
+from tango.step_caches.remote_step_cache import RemoteStepCache
+from tango.step_info import StepInfo, StepState
+from tango.workspace import Run, Workspace
+
+T = TypeVar("T")
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteWorkspace(Workspace):
+    """
+    This is a :class:`~tango.workspace.Workspace` that stores step artifacts on some remote storage location.
+
+    .. tip::
+        All remote workspaces inherit from this.
+    """
+
+    Constants = RemoteConstants
+    STEP_INFO_CACHE_SIZE = 512
+
+    def __init__(self):
+        super().__init__()
+
+        # TODO: remove after conferring.
+        # We make all this into properties because FromParams cannot handle the following construct.
```

Comment:
What does `FromParams` say about the construct?

Reply:
`TypeError: from_params() got an unexpected keyword argument`. Because if it's a keyword argument, `FromParams` cannot tell if it was meant for the parent class, or if it's something used in the constructor of the child class.